server/blob/blob-deduplicating/src/main/scala/GC.scala

```diff
@@ -0,0 +1,196 @@
+}
+
+case class ValidGeneration(id: Long) extends Generation {
+  override def previous: Generation = {
+    if (id == 0) {
+      InvalidGeneration
+    } else {
+      copy(id - 1)
+    }
+  }
+  override def next: Generation = copy(id + 1)
+
+  override def collectibles(targetedGeneration: Generation): Set[Generation] =
+    ((id + 1) to targetedGeneration.previous.previous.asLong).toSet.map(Generation(_))
+
+  override def asLong: Long = id
+}
+
+object InvalidGeneration extends Generation {
+  override def previous: Generation = InvalidGeneration
+
+  override def next: Generation = ValidGeneration(0)
+
+  override def collectibles(targetedGeneration: Generation): Set[Generation] =
+    (0L to targetedGeneration.previous(GC.temporization).asLong).toSet.map(Generation(_))
+
+  override def asLong: Long = -1
+}
+
+object Generation {
+  implicit val ord: Ordering[Generation] = Ordering.by(_.asLong)
+
+  val first: Generation = apply(0)
+
+  def apply(id: Long): Generation = {
+    if (id < 0) {
+      InvalidGeneration
+    } else {
+      ValidGeneration(id)
+    }
+  }
+}
+
+case class Iteration(id: Long, processedGenerations: Set[Generation], lastGeneration: Generation) {
+  def next(generations: Set[Generation], lastGeneration: Generation): Iteration = Iteration(id + 1, generations, lastGeneration)
+}
+
+object Iteration {
+  def initial: Iteration = Iteration(0, Set(InvalidGeneration), InvalidGeneration)
+}
+
+case class ExternalID(id: String) // TODO
+
+sealed trait Event {
+  def blob: BlobId
+  def externalId: ExternalID
+  def generation: Generation
+}
+
+case class Reference(externalId: ExternalID, blobId: BlobId, generation: Generation) extends Event {
+  override def blob: BlobId = blobId
+}
+
+case class Deletion(generation: Generation, reference: Reference) extends Event {
+  override def blob: BlobId = reference.blob
+  override def externalId: ExternalID = reference.externalId
+}
+
+object Events {
+  def getLastGeneration(events: Seq[Event]): Generation = events.map(_.generation).maxOption
+    .getOrElse(Generation.first)
+
+}
+
+case class Report(iteration: Iteration,
+                  blobsToDelete: Set[(Generation, BlobId)]
+                 )
+
+case class State(blobs: Set[BlobId],
+                 references: Map[Generation, Seq[Reference]],
+                 deletions: Map[Generation, Seq[Deletion]]) {
+
+  def stabilize(): StabilizedState = StabilizedState(references, deletions)
+
+  def apply(event: Event): State = { event match {
+    case e : Reference => copy(references = addElement(references, e), blobs = blobs + e.blob)
+    case e : Deletion => copy(deletions = addElement(deletions, e))
+  }
+  }
+
+  private def addElement[T <: Event](collection: Map[Generation, Seq[T]], e: T): Map[Generation, Seq[T]] = {
+    collection.updatedWith(e.generation)(previous => Some(e +: previous.getOrElse(Seq())))
+  }
+}
+
+case class StabilizedState(references: Map[Generation, Seq[Reference]],
```

Comment:
This type is not related to blob store but to the oracle, isn't it?

Reply:
`State` is related to the Oracle, `StabilizedState` to the GC